BEPs/BEP-396.md

```diff
@@ -0,0 +1,187 @@
+
+However, these additional retries waste a lot of computing resources. The read-write set generated by transaction execution can easily build a dependency graph between transactions, called `TxDAG`. This data can be used in the validating block phase to accelerate block execution and further improve the overall performance of the blockchain.
+
+## 5. Specification
+
+In BSC, the block interval is 3s, that is, the more transactions are executed within 3s, the better the chain performance and the greater the throughput.
+
+The two parts that take the most time in the 3s are the validating block and the mining block. Validators often validate the correctness of the current block and continue to mine the next block based on the latest state.
+
+To improve the performance of block execution, transactions need to be executed in parallel as much as possible to make full use of computing and IO resources. Imagine that if the dependencies of transactions are known, then the parallel execution path can be directly determined, but the dependencies need to be obtained by the read-write set after the transaction pre-execution (static analysis is not within the scope of this BEP).
+
+So two different TX execution strategies can be identified:
+
+![](./assets/BEP-396/pic1.png)
+
+In the mining phase, TXs of blocks are executed in parallel or serial to generate `TxDAG`. In the validating phase, `TxDAG` is used to further improve the parallel execution efficiency.
+
+### 5.1 Tx Dependency
+
+Assume that the execution of each transaction is atomic, and the input and output of the execution generate a state read-write set. In concurrent processing, read-only access to the resources is safe, but write operations require exclusive resources.
+
+At the same time, all transactions in the blockchain have a preset order, and transactions with lower TxIndex have higher execution priority.
+
+Therefore, the dependency constraint goes to, if transaction Txi writes `StateA` first, then transaction Txj reads `StateA` and j>i is satisfied. That is to say, Txj depends on Txi, Txj must be executed after Txi.
+
+If a transaction reads and writes the same value at a memory location, the location can be considered read-only.
+
+### 5.2 TxDAG Structure
+
+The structure of `TxDAG` is defined as follows.
+
+```go
+type TxDep struct {
+	// It describes the Relation with below txs 
+	// 0: this tx depends on below txs 
+	// 1: this transaction does not depend on below txs, all other previous txs depend on
+	Relation  uint8
+	TxIndexes []int
+}
+
+type TxDAG struct {
+	// The TxDAG type
+	// 0: delay the distribution of GasFee, it will ignore all gas fee distribution when tx execute
+	// 1: timely distribution of transaction fees, it will keep partial serial execution when tx cannot delay the distribution
+	Type uint8
```

Comment:
The `Type` here will impact the EVM logic(state transition), I think it should not be part of the `TxDAG`,  this should be fixed within the EVM.

Reply:
Because whether or not the Gas Fee is delayed in calculation will result in different TxDAG results, different types are introduced. At the same time, the type field can also be easily used for future expansion.

Reply:
This may lead to state inconsistency. What is the calculation logic of sequential processor, I believe there is only one option to the sequential processor, then for parallel processor, it should adopt the same logic.